### PR TITLE
Protocol2

### PR DESCRIPTION
**Pull request by CaptainHayashi**
Initial support for Protocol2.  Things such as detailed error codes are waiting for spec and not yet implemented.

This is in need of some scrutiny a la @LordAro, because a lot of it is very hacky.

### CODE REVIEW

File: src/audio/audio.hpp [hunk 1]
```diff
@@ -85,14 +85,15 @@ class Audio
...
+	 * example, `Emit("/player", x)` shall be ignored.  This functionality
+	 * is provided in Player.
+	 *
+	 * @param path The path of the response to emit, if possible.
+	 * @param broadcast If true, the emission is an update broadcast.
+	 * @return A pointer to the response, if it exists.
 	 */
-	virtual void Emit(Response::Code code, const ResponseSink *sink,
-	                  size_t id = 0);
+	virtual std::unique_ptr<Response> Emit(const std::string &path, bool broadcast);
```
Comment (LordAro) [inline src/audio/audio.hpp]: Not sure about this returning a unique_ptr - ownership stuff isn't preserved how i'd like - maybe return a raw pointer and capture it in a unique_ptr with the owner object or use a shared_ptr ?

File: src/player.cpp [hunk 2]
```diff
@@ -249,5 +232,95 @@ void Player::SeekRaw(std::uint64_t pos)
...
+	{"/player/time/elapsed", ""}
+};
+
+CommandResult Player::Read(const std::string &path, size_t id) const
+{
+	assert(this->file != nullptr);
+
+	// Maybe the requested item is a directory?
+	auto count = Player::RESOURCES.count(path);
+	if (0 < count) {
```
Comment (LordAro) [inline src/player.cpp]: such yoda conditions

File: src/player.cpp [hunk 3]
```diff
@@ -249,5 +232,95 @@ void Player::SeekRaw(std::uint64_t pos)
...
+	{"/", "/control"},
+	{"/", "/player"},
+	{"/control", "/control/state"},
+	{"/control/state", ""},
+	{"/player", "/player/file"},
+	{"/player", "/player/time"},
+	{"/player/file", ""},
+	{"/player/time", "/player/time/elapsed"},
+	{"/player/time/elapsed", ""}
+};
```
Comment (LordAro) [inline src/player.cpp]: maybe some sort of tree structure would be better?

File: src/player.cpp (hunk 2 above)
Comment (CaptainHayashi) [inline src/player.cpp]: problem you have?

File: src/player.cpp (hunk 3 above)
Comment (CaptainHayashi) [inline src/player.cpp]: maybe, but I just wanted something up quick and dirty with little in the way of code :P

File: src/audio/audio.hpp (hunk 1 above)
Comment (CaptainHayashi) [inline src/audio/audio.hpp]: how do you mean, not preserved?  ownership should be transferring from Emit to the thing that's calling it and staying there, unless I fucked up

**Pull request merged**